decryptor/mysql/prepared_statements.go

```diff
@@ -281,3 +285,103 @@ func (m *mysqlBoundValue) Encode() (encoded []byte, err error) {
 
 	return encoded, outErr
 }
+
+// PreparedStatementFieldTracker track and replace DataType for column and param ColumnDefinition
+type PreparedStatementFieldTracker struct {
+	proxyHandler *Handler
+	// shared value that indicates number of param packet
+	paramsCounter int
+	columnsNum    uint16
+}
+
+// NewPreparedStatementFieldTracker create new PreparedStatementFieldTracker
+func NewPreparedStatementFieldTracker(handler *Handler, columnNum uint16) PreparedStatementFieldTracker {
+	return PreparedStatementFieldTracker{
+		proxyHandler: handler,
+		columnsNum:   columnNum,
+	}
+}
+
+// ParamsTrackHandler ResponseHandler used to track prepared statement params
+func (p *PreparedStatementFieldTracker) ParamsTrackHandler(ctx context.Context, packet *Packet, _, clientConnection net.Conn) error {
+	clientSession := base.ClientSessionFromContext(ctx)
+	if clientSession == nil {
+		p.proxyHandler.logger.Warningln("No packet without ClientSession in context")
+		return nil
+	}
+	items := encryptor.PlaceholderSettingsFromClientSession(clientSession)
+	if items == nil {
+		p.proxyHandler.logger.Debugln("No packet with registered recognized encryption settings")
+		return nil
+	}
+
+	p.proxyHandler.logger.Debugln("Parse param ColumnDefinition")
+	if packet.IsEOF() {
+		p.proxyHandler.resetQueryHandler()
+		// if columns_num > 0 column definition block will follow
+		// https://dev.mysql.com/doc/internals/en/com-stmt-prepare-response.html
+		if p.columnsNum > 0 {
+			p.proxyHandler.setQueryHandler(p.ColumnsTrackHandler)
+		}
+
+		if _, err := clientConnection.Write(packet.Dump()); err != nil {
+			p.proxyHandler.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).
+				Debugln("Can't proxy output")
+		}
+		return nil
+	}
+
+	field, err := ParseResultField(packet)
+	if err != nil {
+		p.proxyHandler.logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorProtocolProcessing).WithError(err).Errorln("Can't parse result field")
+		return err
+	}
+
+	setting, ok := items[p.paramsCounter]
+	if ok {
+		newFieldType, ok := mapEncryptedTypeToField(setting.GetEncryptedDataType())
+		if ok {
+			field.originType = field.Type
+			field.Type = Type(newFieldType)
+			field.changed = true
+		}
+	}
+
+	if _, err := clientConnection.Write(field.Dump()); err != nil {
+		p.proxyHandler.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).
+			Debugln("Can't proxy output")
+	}
+
+	p.paramsCounter++
+	return nil
+}
+
+// ColumnsTrackHandler ResponseHandler used to track prepared statement columns
```

Comment:
```suggestion
// ColumnsTrackHandler used to track prepared statement columns
```

Reply:
ResponseHandler is the separate type - which is special type of function, so ColumnsTrackHandler it the ResponseHandler

Reply:
hm, okay, my fault. Then suggest ot rephrase like `implements ResponseHandler to track prepare statement columns`

Reply:
Makes sense.